### PR TITLE
+htc #18898 modeledCustomHeader to ease matching on headers

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/akka/akka/issues/18898

Though I'm not completely happy yet; the problem is about not having "an additional" type, like now, but "one for all custom headers".
The problem is that custom headers are "abused" (well not really, just used in "special cases" by HTTP internally, and making that compile with the companion approach proved to be very tricky (burnt some time on it today).

Would like some initial feedback, and also opinions if we should keep 2 "custom headers" (I could make the bare one be extended by the Modeled one, and we internally keep using the bare; and tell users to use the Modeled one.

// cc @jrudolph @sirthias @rkuhn

### CODE REVIEW

**Comment by ktoso** [comment]
Whoops test will fail, sorry

**Comment by ktoso** [comment]
Note to self, promised to deliver in M2.

**Comment by ktoso** [comment]
This is now complete (I think – please review) and up for M2.

**Comment by drewhk** [comment]
LGTM I guess

**Comment by ktoso** [comment]
Thanks @drewhk!
Can you take a look at it @sirthias / @jrudolph ?

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 1]
```diff
@@ -57,6 +61,60 @@ abstract class CustomHeader extends jm.headers.CustomHeader {
...
 }
 
+abstract class ModeledCustomHeaderCompanion[H <: ModeledCustomHeader[H]] {
+  def name: String
+  def lowercaseName: String = name.toRootLowerCase
+
+  def parse(value: String): Option[H]
+
+  def apply(value: String): H =
+    parse(value).getOrElse(throw new RuntimeException("Unable to construct custom header by parsing value:" + value))
```
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: I think, as a user, I'd expect a space char here after the `:` in the error message.

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 2]
```diff
@@ -57,6 +61,60 @@ abstract class CustomHeader extends jm.headers.CustomHeader {
...
+//  def name: String = companion.name
+//  def lowercaseName: String = name.toRootLowerCase
+//  final def render[R <: Rendering](r: R): r.type = r ~~ name ~~ ':' ~~ ' ' ~~ value
+//}
+abstract class CustomHeaderCompanion[H <: CustomHeader] {
+  def name: String
+  def lowercaseName: String = name.toRootLowerCase
+
+  def parse(value: String): Option[H]
+  def apply(value: String): H = parse(value).getOrElse(throw new Exception("Whoops"))
```
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: Maybe `RuntimeException` rather than `Exception`?
And an error message that shows the offending input?

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 3]
```diff
@@ -57,6 +61,60 @@ abstract class CustomHeader extends jm.headers.CustomHeader {
...
+  def unapply(raw: RawHeader): Option[H] =
+    if (raw.lowercaseName == lowercaseName) parse(raw.value) else None
+
+  def unapply(value: String): Option[H] =
+    parse(value)
+  //  def unapply(header: H): Option[RawHeader] = Some(RawHeader(header.name, header.value))
```
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: left-over commenting-out?

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 4]
```diff
@@ -57,6 +61,60 @@ abstract class CustomHeader extends jm.headers.CustomHeader {
...
+//abstract class CustomHeader[Self <: CustomHeader[Self]] extends jm.headers.CustomHeader {
+//  def companion: CustomHeaderCompanion[Self]
+//
+//  /** Override to return true if this header shouldn't be rendered */
+//  def suppressRendering: Boolean = false
+//
+//  def name: String = companion.name
+//  def lowercaseName: String = name.toRootLowerCase
+//  final def render[R <: Rendering](r: R): r.type = r ~~ name ~~ ':' ~~ ' ' ~~ value
+//}
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: note to self remove this

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala (hunk 2 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: Ouch, that's a very crappy error message I left here. Thanks for noticing!

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala (hunk 3 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: note to self, remove this

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 5]
```diff
@@ -135,9 +193,21 @@ final case class `If-Range`(entityTagOrDateTime: Either[EntityTag, DateTime]) ex
...
+final class RawHeader(val name: String, val value: String) extends jm.headers.RawHeader {
   val lowercaseName = name.toRootLowerCase
   def render[R <: Rendering](r: R): r.type = r ~~ name ~~ ':' ~~ ' ' ~~ value
+
+  override def equals(obj: Any): Boolean = obj match {
+    case RawHeader(n: Any @unchecked, v) ⇒ (n.toRootLowerCase == lowercaseName) && (v == value)
+    case _                               ⇒ false
+  }
+  override def hashCode(): Int = super.hashCode()
+}
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: I hope I got that right?

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 6]
```diff
@@ -135,9 +193,21 @@ final case class `If-Range`(entityTagOrDateTime: Either[EntityTag, DateTime]) ex
 }
 
 // FIXME: resurrect SSL-Session-Info header once akka.io.SslTlsSupport supports it
-final case class RawHeader(name: String, value: String) extends jm.headers.RawHeader {
+final class RawHeader(val name: String, val value: String) extends jm.headers.RawHeader {
```
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: Do we *have* to drop the `case class` here?
It seems we are loosing the `toString` implementation and `copy` method as well, no?

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: True, thanks. I'll see if possible to keep it a case class and just add the apply/unapply manually.

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: I was able to keep it a case class :+1:

**Comment by ktoso** [comment]
All comments addressed and error reporting improved. RawHeader stays a case class.

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/directives/BasicDirectivesExamplesSpec.scala [hunk 7]
```diff
@@ -233,6 +234,49 @@ class BasicDirectivesExamplesSpec extends RoutingSpec {
       status shouldEqual StatusCodes.BadGateway
     }
   }
+  "1mapResponse-advanced-json" in {
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/directives/BasicDirectivesExamplesSpec.scala]: this was supposed to be a separate commit, removing

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala [hunk 8]
```diff
@@ -138,6 +181,11 @@ final case class `If-Range`(entityTagOrDateTime: Either[EntityTag, DateTime]) ex
 final case class RawHeader(name: String, value: String) extends jm.headers.RawHeader {
   val lowercaseName = name.toRootLowerCase
   def render[R <: Rendering](r: R): r.type = r ~~ name ~~ ':' ~~ ' ' ~~ value
+  override def hashCode(): Int = super.hashCode()
```
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: What's the purpose of this now?

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala]: can be removed, thanks

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/directives/BasicDirectivesExamplesSpec.scala (hunk 7 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/directives/BasicDirectivesExamplesSpec.scala]: (in a separate PR)

**Comment by ktoso** [comment]
Ok, _now_ in a state I'm happy with.

**Comment by sirthias** [comment]
Cool, LGTM now!
